### PR TITLE
CLDR-17177 kbd: update per feedback

### PR DESCRIPTION
**Pull request by srl295**
- documented 'implementation'
- added a Notation section
- corrections to discussion of repository and file layout

CLDR-17177

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by srl295** [comment]
I would like to land #3445 before this, this will need a rebase eventually.

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:267]:
```suggestion
  For example, `id="…keyId"` denotes that `…keyId` (the part between double quotes) is to be replaced with something, in this case a key identifier. As another example, `\u{…usv}` denotes that the `…usv` is to be replaced with something, in this case a Unicode scalar value in hex.
```

I wonder if we should be explicit here -- it's not UTF-16 or other encoding.

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:271]:
```suggestion
When explicitly specified, attribute values can contain escaped characters. This specification uses two methods of escaping, the _UnicodeSet_ notation and the `\u{…usv}` notation.
```

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:318]:
```suggestion
> For distributing keyboards outside of the CLDR repository, it is recommended to either use a copy of the .dtd file, or omit the `DOCTYPE` clause. Do not use a http or https URL in the DOCTYPE.
```

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1006]: Use ellipsis here?

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1037]:
```suggestion
    <display output="\u{0303}" display="◌̃" />  <!-- \u{25CC} \u{0303}-->
    <display output="\u{0303}" display="\u{25cc}\u{0303}" />  <!-- also acceptable -->
```

Given how badly it displays in GitHub preview monospace here, the current example may be counter-productive?

image

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1503]: I think this was original supposed to refer to whitespace characters rather than "spacing marks" (which is kinda weird).

```suggestion
> Strings may contain whitespaces. However, for clarity, it is recommended to escape whitespace characters, even in strings.
```

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:267]: good.. should propagate to other sections also since this is just an example

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1037]: maybe, though it's just an example.

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:136]: Waaay overpromising. The keyboard subcommittee cannot "get device makers, operating systems, apps and services to implement the **script requirements** that Unicode has just added to support their language".

Moreover, any mention of the subcommittees / working groups shouldn't be in our specs. (If there other cases we should remove them also.)

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:228]: This is needlessly speculative. No reason to to that. 
```suggestion
If it becomes necessary in the future, the format could extend the ISO layout to support keys that are located to the left of the "00" column. -->
```

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:287]: why …usv above, and just … here?

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:305]: I could expect to see just sample data in the repository when the spec is in tech preview. Once it is approved, however, I would expect to see real data there (and no sample data).

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:287]:
```suggestion
The `\u{…usv}` notation, a subset of hex notation, is described in [UTS #18 section 1.1](https://www.unicode.org/reports/tr18/#Hex_notation). It can refer to one or multiple individual codepoints. Currently, the following attribute values allow the `\u{…usv}` notation:
```

Ambiguous to me if we should do this but perhaps for consistency

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:753]:
```suggestion
> The `output` attribute contains the sequence of characters that is emitted when pressing this particular key. Control characters, whitespace (other than the regular space character) and combining marks in this attribute are escaped using the `\u{…usv}` notation. More than one key may output the same output.
```

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:312]: "A new mechanism" is way too vague. You might say "before the specification exits tech preview, the intent is to have a different mechanism ..."

The sample data shouldn't use a mechanism that you discourage.

Also, I believe that you allow keyboards to import data: there should be a strong warning about pulling an isolated keyboard out of the repository without everything it depends on, including the dtd.

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:2494]: Just realised this section is all commented out and wrong anyway (e.g. there is no `<map` element. It's a TODO that should be resolved before we publish.

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:1026]: Minor; a list of single line paragraphs is strange. Do you mean this to be a list?

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:2497]: This looks like the element has a prefix "map", like "mapStuff". If you really mean intervening attributes, put a space between map and …

**Review by macchiati — CHANGES_REQUESTED** [review]

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [reply docs/ldml/tr35-keyboards.md:228]: Note that this is in commented out text which should be removed (or fixed)

Comment (mcdurdin) [reply docs/ldml/tr35-keyboards.md:228]: The format no longer uses ISO layout numbers anyway

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [reply docs/ldml/tr35-keyboards.md:2497]: Also in commented out section that needs to be fixed

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:2494]: I'm going to remove the TODO as part of the general cleanup ticket.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:305]: The challenge is that we haven't defined the process for inclusion in the common repository yet. 

The Bengali and Maltese keyboards are real keyboards, not toys. The French azerty keyboard is a test, that could be withdrawn for now.

Perhaps something else here, saying that the process for inclusion is still being defined, and strike the section about sample data?

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:287]: fixed

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md]: Removed this section and removed DOCTYPEs.  (all tests broken, but will fix)

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [reply docs/ldml/tr35-keyboards.md]: Why is this resolved?

Comment (mcdurdin) [reply docs/ldml/tr35-keyboards.md]: The entire commented out section was removed.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:293]: I've updated this to remove the mention of sample data. I'll plan to remove the only keyboard that's only sample data, the azerty one.

Comment (srl295) [reply docs/ldml/tr35-keyboards.md:293]: Mark, I've re-added the DTDs so that sample files pass.  CLDR tooling breaks without a DTD.  Split this out to ticket https://unicode-org.atlassian.net/browse/CLDR-17371

**Comment by srl295** [comment]
@macchiati what's outstanding before this can merge? (The DTD is split off to a separate ticket)

**Review by mcdurdin — DISMISSED** [review]
Latest changes LGTM

**Comment by srl295** [comment]
how did this diverge so? okay..

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:313]: Very soon, this should change...

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:650]: attribute **value** is

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:1312]:
```suggestion
For non-spacing marks, U+25CC `◌` is used as a base. It is an error to use a nonspacing character without a base in the `display` attribute. For example, `display="\u{0303}"` would produce an error.
```

File: docs/ldml/tr35-keyboards.md
Comment (macchiati) [inline docs/ldml/tr35-keyboards.md:1323]: script and languages never prefer anything; they are not animate.

```suggestion
Users of some scripts/languages may prefer a different base than U+25CC. See  [`<displayOptions baseCharacter=…/>`](#element-displayoptions).
```

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:311]: I'll change to point to the keyboard WG page

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:313]: yes. ideally this would merge first.

**Comment by macchiati** [comment]
sounds good

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md]: done

**Review by macchiati — APPROVED** [review]

**Comment by srl295** [comment]
Merge conflicts. Will have to rebase.

**Comment by srl295** [comment]
Rebased with no change. Can I get an approval?

**Review by miloush — APPROVED** [review]

**Pull request merged**
